### PR TITLE
Refactored GeoModel.

### PR DESCRIPTION
**Pull request by davidkellerman**
Changed code in GeoModel to be more "understanding" of GeoJSON structures. It should now produce reasonable results for any valid GeoJSON data.

Changed the `extractGeometryFeatures()` method to set the geometry feature field to a valid GeoJSON object, which is what is needed for the MongoDB `2dsphere` index.

Removed `geometrySingle` property from GeoModel. It wasn't used, and didn't seem to serve a useful purpose. Same for the `extractGeometryFeatures` property.

Removed the `query` parameter from the `findWithin()` and `findIntersects()` methods. It wasn't used, and the code implementing it wasn't correct anyway.

Added `getGeometry()` and `getCentroid()` methods to extract geometry data from the GeoJSON geographic field of a GeoModel instance. Their intended use is for extracting coordinate data that can be used for geo queries. The `getGeometry()` method extracts GeoJSON geometry objects. The `getCentroid()` method calculates the centroid of the geographic field.

Documented the code.

Added tests to check functionality of GeoModel. Wired it up to a `sails-mongo` adapter so it could actually exercise MongoDB geo functionality.

### CODE REVIEW

File: src/GeoModel.js [hunk 1]
```diff
@@ -1,95 +1,290 @@
...
-      acc.push(x)
-    };
-    return acc;
-  }, [])
+function cleanGeometryCoordinates(parts) {
+  let cleaned = []
+  parts.Polygon.forEach((coordinates) => {
+    coordinates = cleanPolygon(coordinates)
+    if (coordinates) {
+      let obj = rewind({type: 'Polygon', coordinates})
```
Comment (ScottMaxson) [inline src/GeoModel.js]: just a question here: are you using rewind as a kind of normalization, or is this about better rendering of geojson on the client? Or both?

File: .gitignore [hunk 2]
```diff
@@ -25,6 +25,5 @@ build/Release
 # Dependency directory
 # https://www.npmjs.org/doc/misc/npm-faq.html#should-i-check-my-node_modules-folder-into-git
 node_modules
-lib
```
Comment (ScottMaxson) [inline .gitignore]: Did you intend to check in 'lib'? Is there a reason I'm missing to put in repo?

File: src/GeoModel.js [hunk 3]
```diff
@@ -1,95 +1,290 @@
...
-  var features = this.geometryFeatureField
-  if (values[orig]) {
-    let val = values[orig]
-    if (_.isString(val)) {
-      val = JSON.parse(val)
+/** Extracts geometry coordinates from a GeoJSON object.
+ * It is strict in that it processes only portions of the object that it
+ * recognizes as valid GeoJSON, but lenient in that it accomodates
+ * missing properties and ignores misplaced or unrecognized properties.
+ * @private
```
Comment (ScottMaxson) [inline src/GeoModel.js]: a few of these are marked @private but ... they seem useful to API?

File: src/GeoModel.js [hunk 4]
```diff
@@ -1,95 +1,290 @@
...
+    }
+  })
+
+  return (geometries.length === 0) ? undefined :
+      (geometries.length > 1) ?
+        { type: 'GeometryCollection', geometries: geometries } :
+        geometries[0]
+}
+
+function extractGeometryFeatures(values, next) {
```
Comment (ScottMaxson) [inline src/GeoModel.js]: doc?

File: src/test/index.js [hunk 5]
```diff
@@ -1,15 +1,68 @@
...
+import {Waterline, BaseModel, GeoModel} from '../'
 
 import sinon from 'sinon'
 import {application as app} from 'nxus-core'
 
 import One from './models/One'
 import Two from './models/Two'
 
+// example from RFC 7946, section 1.5 (deconstructed and reassembled)
+const geoJSONPoint = { 'type': 'Point', 'coordinates': [102.0, 0.5] },
```
Comment (ScottMaxson) [inline src/test/index.js]: I do like UPPER_CAMEL for constants. But maybe you don't? Just a thought.

File: src/GeoModel.js (hunk 1 above)
Comment (davidkellerman) [inline src/GeoModel.js]: Yes, normalization. The idea is for the geometry features field to be "trustworthy" – valid GeoJSON, just GeoJSON geometry objects, consistent organization. The incoming geometry field can have all sorts of quirks. The code does a few things to ensure the Polygon objects are valid GeoJSON – discards duplicate points, ensures closed paths, ensures clockwise winding order.

File: .gitignore (hunk 2 above)
Comment (davidkellerman) [inline .gitignore]: This from one of the commit comments (probably not at all obvious):

>  Add lib/ files to repository.
>    
> Allows npm to install from GitHub repoisitory. Useful for PR branch; we can remove it before merging back into master.

File: src/GeoModel.js (hunk 3 above)
Comment (davidkellerman) [inline src/GeoModel.js]: I think you could be right – some of these _might_ be useful as static methods. However, my bias is to leave them private until we have a demonstrated need.

File: src/test/index.js (hunk 5 above)
Comment (davidkellerman) [inline src/test/index.js]: That was my bias back in the C/C++ and Java days. Now, I'm not so sure the distinction is worth highlighting; so have tended towards declaring as `const` where appropriate, but using camel-case everywhere.

**Pull request merged**